hypothesis-python/src/hypothesis/internal/conjecture/dfa/lstar.py

```diff
@@ -0,0 +1,222 @@
+   usually most strings will be mispredicted near the beginning.
+
+A note on performance: This code is not really fast enough for
+us to ever want to run in production on large strings, and this
+is somewhat intrinsic. We should only use it in testing or for
+learning languages offline that we can record for later use.
+
+"""
+
+
+class LStar:
+    def __init__(self, member):
+        self.__experiments = []
+        self.__cache = {}
+        self.__member = member
+
+        self.__add_experiment(b"")
+
+    def member(self, s):
+        """Check whether this string is a member of the language
+        to be learned."""
+        s = bytes(s)
+        try:
+            return self.__cache[s]
+        except KeyError:
+            return self.__cache.setdefault(s, self.__member(s))
+
+    @property
+    def generation(self):
+        """Return an integer value that will be incremented
+        every time the DFA we predict changes."""
+        return len(self.__experiments)
+
+    @property
+    def dfa(self):
+        """Returns our current model of a DFA for matching
+        the language we are learning."""
+        if self.__dfa is None:
+            self.__dfa = ExperimentDFA(self.member, self.__experiments)
+        return self.__dfa
+
+    def learn(self, s):
+        """Learn to give the correct answer on this string.
+        That is, after this method completes we will have
+        ``self.dfa.matches(s) == self.member(s)``.
+
+        Note that we do not guarantee that this will remain
+        true in the event that learn is called again with
+        a different string. It is in principle possible that
+        future learning will cause us to make a mistake on
+        this string. However, repeatedly calling learn on
+        each of a set of strings until the generation stops
+        changing is guaranteed to terminate.
+        """
+        s = bytes(s)
+        correct_outcome = self.member(s)
+
+        # We don't want to check this inside the loop because it potentially
+        # causes us to evaluate more of the states than we actually need to,
+        # but if our model is mostly correct then this will be faster because
+        # we only need to evaluate strings that are of the form
+        # ``state + experiment``, which will generally be cached and/or needed
+        # later.
+        if self.dfa.matches(s) == correct_outcome:
+            return
+
+        # In the papers they assume that we only run this process
+        # once, but this is silly - often when you've got a messy
+        # string it will be wrong for many different reasons.
+        #
+        # Thus we iterate this to a fixed point where we repair
+        # the DFA by repeatedly adding experiments until the DFA
+        # agrees with the membership function on this string.
+        while True:
+            dfa = self.dfa
+
+            states = [dfa.start]
+
+            def accept(n):
```

Comment:
I find the name “accept” confusing here, because it’s not the same concept as the DFA accepting a string.

(Maybe something along the lines of “understands”?)

Reply:
Fair! `accept` is basically my default "this function is too long for a lambda" name but it's confusing in context, you're right.

I went with `seems_right` because that's more or less what the function checks.